examples/all-clusters-app/all-clusters-common/src/EnergyEvseDelegateImpl.cpp

```diff
@@ -0,0 +1,771 @@
+    {
+        return Status::ConstraintError;
+    }
+
+    /* there is no attribute to store this so store in private variable */
+    mCableAssemblyCurrentLimit = currentmA;
+
+    return this->ComputeMaxChargeCurrentLimit();
+}
+
+/**
+ * @brief    Called by EVSE Hardware to indicate a fault
+ *
+ * @param    StateEnum
+ */
+Status EnergyEvseDelegate::HwSetState(StateEnum state)
+{
+    /* the only allowed states that the EVSE hardware can set are:
+     *  kNotPluggedIn
+     *  kPluggedInNoDemand
+     *  kPluggedInDemand
+     */
+    switch (state)
+    {
+    case StateEnum::kNotPluggedIn:
+        // TODO - work out logic here
+        mHwState = state;
+        break;
+    case StateEnum::kPluggedInNoDemand:
+        // TODO - work out logic here
+        mHwState = state;
+        break;
+    case StateEnum::kPluggedInDemand:
+        // TODO - work out logic here
+        mHwState = state;
+        break;
+
+    default:
+        /* All other states should be managed by the Delegate */
+        // TODO (assert?)
+        break;
+    }
+
+    return Status::Success;
+}
+
+/**
+ * @brief    Called by EVSE Hardware to indicate a fault
+ *
+ * @param    FaultStateEnum
+ */
+Status EnergyEvseDelegate::HwSetFault(FaultStateEnum fault)
+{
+    ChipLogProgress(AppServer, "EnergyEvseDelegate::Fault()");
+
+    if (fault == FaultStateEnum::kNoError)
+    {
+        /* Update State to previous state */
+        // TODO: need to work out the logic here!
+
+        /* Update SupplyState to previous state */
+    }
+    else
+    {
+        /* Update State & SupplyState */
+        SetState(StateEnum::kFault);
+        SetSupplyState(SupplyStateEnum::kDisabledError);
+    }
+
+    /* Update FaultState */
+    SetFaultState(fault);
+
+    // TODO: Generate events
+
+    return Status::Success;
+}
+
+/**
+ * @brief    Called by EVSE Hardware to Send a RFID event
+ *
+ * @param    ByteSpan RFID tag value (max 10 octets)
+ */
+Status EnergyEvseDelegate::HwSetRFID(ByteSpan uid)
+{
+    Events::Rfid::Type event{ .uid = uid };
+    EventNumber eventNumber;
+    CHIP_ERROR error = LogEvent(event, mEndpointId, eventNumber);
+    if (CHIP_NO_ERROR != error)
+    {
+        ChipLogError(Zcl, "[Notify] Unable to send notify event: %s [endpointId=%d]", error.AsString(), mEndpointId);
+        return Status::Failure;
+    }
+
+    return Status::Success;
+}
+/**
+ * @brief    Called by EVSE Hardware to share the VehicleID
+ *
+ * This routine will make a copy of the string so the callee doesn't
+ * have to hold onto it forever.
+ *
+ * @param    CharSpan containing up to 32 chars.
+ */
+Status EnergyEvseDelegate::HwSetVehicleID(const CharSpan & newValue)
+{
+    if (mVehicleID.IsNull() || !newValue.data_equal(mVehicleID.Value()))
+    {
+        /* create a copy of the string so the callee doesn't have to keep it */
+        char * destinationBuffer = new char[kMaxVehicleIDBufSize];
```

Comment:
use RAII here. ScopedBuffer classes exist to ensure it handles resource cleanup for you.

Reply:
As discussed on slack this may not be clear - but the memory is alloc'd here, and held until the next time the VehicleID is updated.

Then the old value is deleted and  replaced with the new value:
```
    if (!mVehicleID.IsNull())
    {
        delete[] mVehicleID.Value().data();
    }
    mVehicleID = MakeNullable(static_cast<CharSpan>(destinationString)); 
```

Per the discussion - there was a potential leak when the Delegate is destroyed. So have pushed an update where the destructor deletes this.

Needs refactoring as part of https://github.com/project-chip/connectedhomeip/issues/30993